pkg/kvstore/allocator/cache.go

```diff
@@ -0,0 +1,280 @@
+// Copyright 2016-2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package allocator
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/lock"
+
+	"github.com/sirupsen/logrus"
+)
+
+// idMap provides mapping from ID to an AllocatorKey
+type idMap map[ID]AllocatorKey
+
+// keyMap provides mapping from AllocatorKey to ID
+type keyMap map[string]ID
+
+type cache struct {
+	backend  kvstore.BackendOperations
+	prefix   string
+	stopChan chan bool
+
+	// mutex protects all cache data structures
+	mutex lock.RWMutex
+
+	// cache is a local cache of all IDs allocated in the kvstore. It is
+	// being maintained by watching for kvstore events and can thus lag
+	// behind.
+	cache idMap
+
+	// keyCache shadows cache and allows access by key
+	keyCache keyMap
+
+	// nextCache is the cache is constantly being filled by startWatch(),
+	// when startWatch has successfully performed the initial fill using
+	// ListPrefix, the cache above will be pointed to nextCache. If the
+	// startWatch() fails to perform the initial list, then the cache is
+	// never pointed to nextCache. This guarantees that a valid cache is
+	// kept at all times.
+	nextCache idMap
+
+	// nextKeyCache follows the same logic as nextCache but for keyCache
+	nextKeyCache keyMap
+
+	stopWatchWg sync.WaitGroup
```

Comment:
Can you add a comment for this variable?

Reply:
done